crates/common/mqtt_channel/src/errors.rs

```diff
@@ -90,3 +94,37 @@ impl From<futures::channel::mpsc::SendError> for MqttError {
         MqttError::SendOnClosedConnection
     }
 }
+
+pub fn from_connection_error(err: rumqttc::ConnectionError) -> Result<(), MqttError> {
```

Comment:
I would wrap this as an `MqttError` function so this will be invoked as `MqttError::from_connection_error` (otherwise the name sounds incomplete: *what* is built from the connection error?).

The type signature is also unusual for a *from* function. I read this as building a unit value with conversion error that are `MqttError`. While what you meant is building an `MqttError`.

```suggestion
impl MqttError {
   pub fn from_connection_error(err: rumqttc::ConnectionError) -> MqttError {
```


I agree that there is a wart to address. This is the case of `ConnectionError::ConnectionRefused(ConnectReturnCode::Success)`. I would simply treat this case as an error in the `from_connection_error` function and check this specific case in the `init_session` function.